stablehlo/dialect/TypeInference.cpp

```diff
@@ -1223,7 +1223,35 @@ static LogicalResult inferGatherReturnTypeComponents(
                             offsetDims, collapsedSliceDims, startIndexMap,
                             indexVectorDim, shape);
 
-  inferredReturnShapes.emplace_back(shape, elementType);
+  // The dimension sizes of result, corresponding to offset dimensions, depend
+  // on attributes (like `collapsed_slice_dims` and `slice_sizes`) and hence are
+  // always static. Whereas, the dimension sizes of result, corresponding to
+  // batch dimensions, depends on input `start_indices` and could be dynamic.
+  // The corresponding bounds, in that case,  are propagated from the
+  // `start_indices`.
+  ArrayRef<int64_t> startIndicesBounds = encodingToBounds(
+      startIndices.getType().cast<RankedTensorType>().getEncoding());
+  SmallVector<int64_t> inferredBounds(resultRank, ShapedType::kDynamic);
+  if (!startIndicesBounds.empty()) {
+    SmallVector<int64_t> batchDims;
+    for (int dim = 0; dim < resultRank; ++dim)
+      if (!llvm::is_contained(offsetDims, dim)) batchDims.push_back(dim);
+
+    for (int i = 0; i < resultRank; ++i) {
+      auto* batchDimsIt = std::find(batchDims.begin(), batchDims.end(), i);
+      if (batchDimsIt == batchDims.end()) continue;
```

Comment:
for non batch dims, there is this handling in XLA.
https://github.com/tensorflow/tensorflow/blob/85d9f5aef28d6a0aaea6aff0b261e11254f9320c/tensorflow/compiler/xla/service/shape_inference.cc#L3572

Current StableHLO spec with static doesn't allow slice sizes to be larger than the input but we need to figure out semantics in dynamism RFC. Either follow the XLA semantics (as per the implementation) or disallow that. Check with Eugene on what to do here until then.

Reply:
That is a good point. Thanks for pointing! 
So, if the size sizes is larger than input, then then the bounds of input should be propagated to result's bounds.   

cc @burmako Can you please help here?

Reply:
That's a nice subtlety - thank you for calling this out! Let's keep the current implementation (i.e. what's in the PR) for now, and I'll make a note for the dynamism RFC.